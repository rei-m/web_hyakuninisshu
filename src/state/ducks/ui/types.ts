import { Action } from 'redux';
import { Color, Kimariji } from '@src/types';
import * as constants from './constants';

export type OPEN_KARUTAS_FILTER_TYPE = typeof constants.OPEN_KARUTAS_FILTER_NAME;

export type CLOSE_KARUTAS_FILTER_TYPE = typeof constants.CLOSE_KARUTAS_FILTER_NAME;

export type TOGGLE_KARUTAS_COLOR_TYPE = typeof constants.TOGGLE_KARUTAS_COLOR_NAME;

export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export interface KarutasFilter {
  readonly colors: Array<{ color: Color; checked: boolean }>;
  readonly kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}

export interface UiState {
  readonly karutasFilter: KarutasFilter & { readonly open: boolean };
}
