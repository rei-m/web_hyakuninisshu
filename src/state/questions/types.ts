import { Action } from 'redux';
import { Kimariji, Color, QuestionId, KarutaNo, Karuta } from '@src/domain/models';
import * as constants from './constants';
import { Meta, Payload } from '..';

type Condition<T> = {
  values: Array<T>;
};

export type RangeFromCondition = 1 | 11 | 21 | 31 | 41 | 51 | 61 | 71 | 81 | 91;
export const RangeFromCondition: Condition<RangeFromCondition> = {
  values: [1, 11, 21, 31, 41, 51, 61, 71, 81, 91],
};

export type RangeToCondition = 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 100;
export const RangeToCondition: Condition<RangeToCondition> = {
  values: [10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
};

export type KimarijiCondition = null | Kimariji;
export const KimarijiCondition: Condition<KimarijiCondition> = {
  values: [null, 1, 2, 3, 4, 5, 6],
};

export type ColorCondition = null | Color;
export const ColorCondition: Condition<ColorCondition> = {
  values: [null, 'blue', 'pink', 'yellow', 'green', 'orange'],
};

export type KarutaStyleCondition = 'kanji' | 'kana';
export const KarutaStyleCondition: Condition<KarutaStyleCondition> = {
  values: ['kanji', 'kana'],
};

export type QuestionAnimCondition = 'none' | 'slow' | 'normal' | 'fast';
export const QuestionAnimCondition: Condition<QuestionAnimCondition> & {
  toDulation: (value: QuestionAnimCondition) => number;
} = {
  values: ['none', 'slow', 'normal', 'fast'],
  toDulation: (value: QuestionAnimCondition) => {
    switch (value) {
      case 'normal':
        return 0.6;
      case 'fast':
        return 0.3;
      case 'slow':
        return 1.0;
      default:
        return 0;
    }
  },
};

export type YomiFuda = Readonly<{
  karutaNo: KarutaNo;
  shoku: string;
  niku: string;
  sanku: string;
}>;

export type ToriFuda = Readonly<{
  karutaNo: KarutaNo;
  shiku: string;
  kekku: string;
}>;

export type StartTrainingAction = Action<typeof constants.START_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }> &
  Meta<{
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  }>;

export type RestartTrainingAction = Action<typeof constants.RESTART_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartExamAction = Action<typeof constants.START_EXAM_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartQuestionAction = Action<typeof constants.START_QUESTION_NAME> &
  Payload<{
    questionId: QuestionId;
    content: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
  }>;

export type ResetQuestionAction = Action<typeof constants.RESET_QUESTION_NAME>;

export type FinishQuestionAction = Action<typeof constants.FINISH_QUESTION_NAME> &
  Payload<{
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  }>;

export type AnswerQuestionAction = Action<typeof constants.ANSWER_QUESTION_NAME> &
  Payload<{
    isCorrect: boolean;
    selectedKarutaNo: KarutaNo;
    correctKaruta: Karuta;
  }>;

export type ConfirmCorrectAction = Action<typeof constants.CONFIRM_CORRECT_NAME>;

export type OpenNextQuestionAction = Action<typeof constants.OPEN_NEXT_QUESTION_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
  }>;

export type Actions =
  | StartTrainingAction
  | RestartTrainingAction
  | StartExamAction
  | StartQuestionAction
  | ResetQuestionAction
  | FinishQuestionAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction;

export type ActionCreator = {
  /**
   * 指定した条件から問題を作成/保存してStartTrainingActionを返す
   *
   * @param rangeFrom 出題範囲From
   * @param rangeTo 出題範囲To
   * @param kimariji 決まり字条件
   * @param color 色条件
   * @param kamiNoKuStyle 上の句表示形式
   * @param shimoNoKuStyle 下の句表示形式
   * @param questionAnim 問題の表示速度
   *
   * @returns {StartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startTraining: (
    rangeFrom: RangeFromCondition,
    rangeTo: RangeToCondition,
    kimariji: KimarijiCondition,
    color: ColorCondition,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    questionAnim: QuestionAnimCondition
  ) => StartTrainingAction;

  /**
   * 回答済の問題の中から間違えた問題をもとに問題を作成/保存してRestartTrainingActionを返す
   *
   * @returns {RestartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  restartTraining: () => RestartTrainingAction;

  /**
   * 全ての歌を対象に問題を作成/保存してStartExamActionを返す
   *
   * @returns {StartExamAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startExam: () => StartExamAction;

  /**
   * 問題の回答を開始してStartQuestionActionを返す
   *
   * @returns {StartQuestionAction}
   * @throws {Error} 指定した問題がない場合
   */
  startQuestion: (
    currentQuestionId: QuestionId,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    startDate: Date
  ) => StartQuestionAction;

  /**
   * 問題の回答を判定してAnswerQuestionActionを返す
   *
   * @returns {AnswerQuestionAction}
   * @throws {Error} 指定した問題がない場合
   * @throws {Error} 問題の回答が開始されていない場合
   */
  answerQuestion: (currentQuestionId: QuestionId, toriFuda: ToriFuda, answerDate: Date) => AnswerQuestionAction;

  /**
   * ConfirmCorrectActionを返す
   *
   * @returns {ConfirmCorrectAction}
   */
  confirmCorrect: () => ConfirmCorrectAction;

  /**
   * 次の問題を取得してOpenNextQuestionActionを返す
   *
   * @returns {OpenNextQuestionAction}
   * @throws {Error} 次の問題がない場合
   */
  openNextQuestionAction: (currentQuestionId: QuestionId) => OpenNextQuestionAction;

  /**
   * 問題を初期状態に戻してResetQuestionActionを返す
   *
   * @returns {ResetQuestionAction}
   */
  resetQuestion: () => ResetQuestionAction;

  /**
   * 問題を終了して結果を見る
   *
   * @returns {FinishQuestionAction}
   */
  finishQuestion: () => FinishQuestionAction;
};

export type State = Readonly<{
  state: 'waiting' | 'ready' | 'playing' | 'confirm' | 'finished';
  totalCount: number;
  currentPosition: number;
  currentQuestion?: {
    questionId: QuestionId;
    content?: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
    answer?: {
      isCorrect: boolean;
      selectedKarutaNo: KarutaNo;
      correctKaruta: Karuta;
    };
  };
  result?: {
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  };
  trainingCondition: {
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  };
}>;
