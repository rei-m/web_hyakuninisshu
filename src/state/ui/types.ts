import { Action } from 'redux';
import * as constants from './constants';
import { Color, Kimariji } from '@src/domain/models';
import { Payload } from '..';

export type KarutasFilter = Readonly<{
  colors: Array<{ color: Color; checked: boolean }>;
  kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}>;

export type OpenKarutasFilterAction = Action<typeof constants.OPEN_KARUTAS_FILTER_NAME>;

export type CloseKarutasFilterAction = Action<typeof constants.CLOSE_KARUTAS_FILTER_NAME>;

export type ToggleKarutasColorAction = Action<typeof constants.TOGGLE_KARUTAS_COLOR_NAME> &
  Payload<{
    color: Color;
    checked: boolean;
  }>;

export type ToggleKarutasKimarijiAction = Action<typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME> &
  Payload<{
    kimariji: Kimariji;
    checked: boolean;
  }>;

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export type ActionCreator = {
  /**
   * OpenKarutasFilterActionを返す
   *
   * @returns {OpenKarutasFilterAction}
   */
  openKarutasFilter: () => OpenKarutasFilterAction;

  /**
   * CloseKarutasFilterActionを返す
   *
   * @returns {CloseKarutasFilterAction}
   */
  closeKarutasFilter: () => CloseKarutasFilterAction;

  /**
   * ToggleKarutasColorActionを返す
   *
   * @param color 歌の色
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasColorAction}
   */
  toggleKarutasColor: (color: Color, checked: boolean) => ToggleKarutasColorAction;

  /**
   * ToggleKarutasKimarijiActionを返す
   *
   * @param kimariji 歌の決まり字
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasKimarijiAction}
   */
  toggleKarutasKimariji: (kimariji: Kimariji, checked: boolean) => ToggleKarutasKimarijiAction;
};

export type State = Readonly<{
  karutasFilter: KarutasFilter & { open: boolean };
}>;
