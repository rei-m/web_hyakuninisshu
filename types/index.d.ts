declare module '*.png'
declare module '*.jpg'

// http://ideasintosoftware.com/typescript-advanced-tricks/s
type Diff<T extends string, U extends string> = ({[P in T]: P } & {[P in U]: never } & { [x: string]: never })[T];  
type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>; 
