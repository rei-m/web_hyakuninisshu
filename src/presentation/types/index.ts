import { RouteComponentProps } from '@reach/router';

export type SiteMetaData = {
  title: string;
  description: string;
  author: string;
};

export type GeneratedPageComponentProps<T = {}> = { pageContext: T } & RouteComponentProps<T>;

export type AppErrorType = 'Network' | 'NotFound' | 'Unknown';

export class AppError implements Error {
  private _message: string;
  private _type: AppErrorType;

  constructor(message: string, type: AppErrorType) {
    this._message = message;
    this._type = type;
  }

  public get name() {
    return 'AppError';
  }

  public get message() {
    return this._message;
  }

  public get type() {
    return this._type;
  }

  public toString() {
    return `${this.name}: ${this.message}`;
  }
}
