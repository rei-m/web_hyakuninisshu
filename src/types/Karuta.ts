interface Karuta {
  id: number;
  imageNo: string;
}

export default Karuta;
