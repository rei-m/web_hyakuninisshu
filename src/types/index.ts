export interface Karuta {
  id: number;
  imageNo: string;
}

export interface Question {
  correctKaruta: Karuta;
  choices: Karuta[];
}

export interface OptionItem<T> {
  readonly name: string;
  readonly value: T;
}
