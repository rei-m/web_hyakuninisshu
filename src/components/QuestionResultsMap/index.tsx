import * as React from 'react';
import styled from 'styled-components';
import { withAppTheme } from '../../styles';
import { Answer, Karuta, Question } from '../../types';
import { toKarutaIdString } from '../helper';
import * as correctImage from './check_correct.png';
import * as incorrectImage from './check_incorrect.png';

export interface QuestionResultsMapProps {
  readonly questions: Question[];
  readonly answers: Answer[];
  readonly style?: React.CSSProperties;
  readonly onClickResult: (karuta: Karuta) => void;
}

export interface QuestionResultsCellProps {
  readonly question: Question;
  readonly answer: Answer;
  readonly onClickResult: (karuta: Karuta) => void;
}

const Root = withAppTheme(styled.div)`
  display: flex;
  flex-wrap: wrap;
  background-color: ${({ theme }) => theme.colorThin};
  box-sizing: border-box;
  box-shadow: ${({ theme }) => theme.elevationShadow1x};
  border-radius: 4px;
`;

const Cell = styled.div`
  width: 20%;
  height: 69px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const KarutaNo = styled.span`
  font-size: 1.2rem;
`;

const CorrentImageBox = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
`;

export const QuestionResultsCell = ({
  question,
  answer,
  onClickResult
}: QuestionResultsCellProps) => {
  const onClickCell = () => {
    onClickResult(question.correctKaruta);
  };
  return (
    <Cell onClick={onClickCell}>
      <KarutaNo>{toKarutaIdString(question.correctKaruta.id)}</KarutaNo>
      <CorrentImageBox>
        <img
          src={answer.correct ? correctImage : incorrectImage}
          style={{ width: '80%' }}
        />
      </CorrentImageBox>
    </Cell>
  );
};

const QuestionResultsMap = ({
  questions,
  answers,
  style,
  onClickResult
}: QuestionResultsMapProps) => (
  <Root style={style}>
    {questions.map((q, i) => (
      <QuestionResultsCell
        question={q}
        answer={answers[i]}
        onClickResult={onClickResult}
        key={i}
      />
    ))}
  </Root>
);

export default QuestionResultsMap;
