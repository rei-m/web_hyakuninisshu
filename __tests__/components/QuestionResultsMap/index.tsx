import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import { create } from '../../factories';
import { sel } from '../../helpers';
import QuestionResultsMap, {
  QuestionResultsMapProps
} from '@src/components/QuestionResultsMap';
import { Answer, Question } from '@src/types';

describe('<QuestionResultsMap />', () => {
  let baseProps: QuestionResultsMapProps;

  beforeEach(() => {
    const questions = [
      create<Question>('question'),
      create<Question>('question')
    ];

    const answers = [
      create<Answer>('answer', {
        questionId: questions[0].id
      }),
      create<Answer>('answer', {
        questionId: questions[1].id
      })
    ];

    baseProps = {
      answers,
      onClickResult: jest.fn(),
      questions
    };
  });

  it('should render component', () => {
    const renderer = ReactTestRenderer.create(
      <QuestionResultsMap {...baseProps} />
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClickResult when result clicked', () => {
    const wrapper = shallow(<QuestionResultsMap {...baseProps} />);
    wrapper
      .find(sel(`question-${baseProps.questions[0].id}`))
      .simulate('click');
    expect(baseProps.onClickResult).toHaveBeenCalledWith(
      baseProps.questions[0].correctKaruta
    );
  });
});
