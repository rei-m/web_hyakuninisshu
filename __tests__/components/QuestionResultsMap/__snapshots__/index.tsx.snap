// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionResultsMap /> should render component 1`] = `
<div
  className="sc-bdVaJa biXiia"
  style={undefined}
>
  <div
    className="sc-bwzfXH igTYpN"
    data-test="question-1"
    onClick={[Function]}
  >
    <span
      className="sc-htpNat buCuQu"
    >
      一番
    </span>
    <span
      className="sc-bxivhb hSyinz"
    >
      <img
        src="check_correct.png"
        style={
          Object {
            "width": "80%",
          }
        }
      />
    </span>
  </div>
  <div
    className="sc-bwzfXH igTYpN"
    data-test="question-2"
    onClick={[Function]}
  >
    <span
      className="sc-htpNat buCuQu"
    >
      一番
    </span>
    <span
      className="sc-bxivhb hSyinz"
    >
      <img
        src="check_correct.png"
        style={
          Object {
            "width": "80%",
          }
        }
      />
    </span>
  </div>
</div>
`;
